src/ray/gcs/client_test.cc

```diff
@@ -642,6 +642,81 @@ TEST_F(TestGcsWithAsio, TestLogSubscribeAll) {
   TestLogSubscribeAll(job_id_, client_);
 }
 
+void TestSetSubscribeAll(const JobID &job_id,
+                         std::shared_ptr<gcs::AsyncGcsClient> client) {
+  std::vector<ObjectID> object_ids;
+  for (int i = 0; i < 3; i++) {
+    object_ids.emplace_back(ObjectID::from_random());
+  }
+  std::vector<std::string> managers = {"abc", "def", "ghi"};
+
+  // Callback for a notification.
+  auto notification_callback = [object_ids, managers](
+                                   gcs::AsyncGcsClient *client, const UniqueID &id,
+                                   const GcsTableNotificationMode mode,
+                                   const std::vector<ObjectTableDataT> data) {
+    if (test->NumCallbacks() < 3 * 3) {
+      ASSERT_EQ(mode, GcsTableNotificationMode::APPEND_OR_ADD);
+    } else {
+      ASSERT_EQ(mode, GcsTableNotificationMode::REMOVE);
+    }
+    ASSERT_EQ(id, object_ids[test->NumCallbacks() / 3 % 3]);
+    // Check that we get notifications in the same order as the writes.
+    for (const auto &entry : data) {
+      ASSERT_EQ(entry.manager, managers[test->NumCallbacks() % 3]);
+      test->IncrementNumCallbacks();
+    }
+    if (test->NumCallbacks() == object_ids.size() * 3 * 2) {
+      test->Stop();
+    }
+  };
+
+  // Callback for subscription success. We are guaranteed to receive
+  // notifications after this is called.
+  auto subscribe_callback = [job_id, object_ids, managers](gcs::AsyncGcsClient *client) {
+    // We have subscribed. Do the writes to the table.
+    for (size_t i = 0; i < object_ids.size(); i++) {
+      for (size_t j = 0; j < managers.size(); j++) {
+        auto data = std::make_shared<ObjectTableDataT>();
+        data->manager = managers[j];
+        for (int k = 0; k < 3; k++) {
+          // Add the same entry several times.
+          // Expect no notification if the entry already exists.
+          RAY_CHECK_OK(client->object_table().Add(job_id, object_ids[i], data, nullptr));
+        }
+      }
+    }
+    for (size_t i = 0; i < object_ids.size(); i++) {
+      for (size_t j = 0; j < managers.size(); j++) {
+        auto data = std::make_shared<ObjectTableDataT>();
+        data->manager = managers[j];
+        for (int k = 0; k < 3; k++) {
+          // Remove the same entry several times.
+          // Expect no notification if the entry doesn't exist.
+          RAY_CHECK_OK(client->object_table().Remove(job_id, object_ids[i], data, nullptr));
```

Comment:
I was under the impression that `ObjectTable::Remove` would be fatal if the object ID to remove doesn't exist, since we check for Redis errors [here](https://github.com/ray-project/ray/blob/70d2bd6612b85bdffd292d8e4ec7d985df7c5909/src/ray/gcs/redis_context.cc#L58). Is that not the case?

Reply:
In our internal codebase, return OK for removing a non-existent entry is necessary because we need to make sure the operation is idempotent. GCS operations must be idempotent because we deal with redis reconnection and failed commands may be resent after redis failover.

However, return OK for removing non-existent entry can be buggy because the caller may try to remove an entry with slightly different field values. (it's usual in current codebase because we have a special field to mark its write mode, so we often only fill key fields such as ids if we want to append an entry with deletion mode) This results in unexpected behavior and hard to find the bug.

I'll change `RAY.SET_REMOVE` to return an error if the entry doesn't exist. And keep `RAY.SET_ADD` to return OK if the entry already exists. Sounds OK?